hashicorp/dev-portal
# feat: toast component & learn opt-in welcome toast

## ✅ "Ready for Review" checklist


---

## 🔗 Relevant links

- [Preview - `/waypoint/tutorials?optInFrom=learn`][preview-tutorials-waypoint] 🔎
- [Preview - `/vault/tutorials?optInFrom=learn`][preview-tutorials-vault] 🔎
- [Preview - `/swingset/components/toast`][preview-swingset] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

Implements `components/toast`, and uses it to welcome visitors who have just opted into the beta.
## 🤷 Why

We want to surface a toast notification to visitors when they first opt-in to the beta.

## 🛠️ How

- Implements toast components with [react-hot-toast]
    - Implementation is very similar to [@hashicorp/react-toast], and also draws inspiration from toasts in [hashicorp/learn], both of which also use [react-hot-toast]
    - Aims to match [HDS Toast component][hds-toast] 
- Uses `components/toast` to implement an opt-in toast to welcome visitors to the beta

## 📸 Design Screenshots

<img width="1162" alt="CleanShot 2022-05-22 at 23 20 03@2x" src="https://user-images.githubusercontent.com/4624598/169736958-6f747abe-114e-4a34-9dcf-648f110829c7.png">

<img width="1440" alt="toast-stickersheet" src="https://user-images.githubusercontent.com/4624598/169442929-5551c3b7-1109-4929-8cf9-e7021bee7d0f.png">

## 🧪 Testing

- Visit any tutorials page, such as [/waypoint/tutorials?optInFrom=learn][preview-tutorials-waypoint]
    - You should see a "Welcome" toast pop up
    - This toast should be announced with `aria-live="polite"`. The action buttons on the toast should not be announced.
- The behaviour above should be consistent across tutorial URLs with an `?optInFrom` parameter
    - Product tutorial page - other page is [/vault/tutorials?optInFrom=learn][preview-tutorials-vault]
    - Collection page - [/waypoint/tutorials/get-started-docker?optInFrom=learn](https://dev-portal-git-zstoast-components-hashicorp.vercel.app/waypoint/tutorials/get-started-docker?optInFrom=learn)
    - Tutorial page - for example [/waypoint/tutorials/get-started-docker/get-started-intro?optInFrom=learn](https://dev-portal-git-zstoast-components-hashicorp.vercel.app/waypoint/tutorials/get-started-docker/get-started-intro?optInFrom=learn)
- Visit [the Swingset page for Toast][preview-swingset] and confirm the component looks to spec
    - Some small elements, eg the `×`-close-button, were hard to track down in spec, so I matched what I could see at the [HDS Toast Component page][hds-toast]

## 💭 Anything else?

- Intent is to build in this PR what we need for toasts for June 20
    - Apart from the downloads view `code-block` toast in this PR, I think this is only opt-in toasts
- Intent is _not_  to fully match the [HDS Toast Component][hds-toast] spec, _yet_
    - For example, arbitrary content such as italics, bold, and inline links are not _yet_ supported in the `description` prop. Intent is to implement that functionality later, if we need it. (May be sooner rather than later, I did see some design iterations that use inline links within the opt-in toasts, but I believe those are not the most current).

[preview-tutorials-waypoint]: https://dev-portal-git-zstoast-components-hashicorp.vercel.app/waypoint/tutorials?optInFrom=learn
[preview-tutorials-vault]: https://dev-portal-git-zstoast-components-hashicorp.vercel.app/vault/tutorials?optInFrom=learn
[preview-downloads]: https://dev-portal-git-zstoast-components-hashicorp.vercel.app/waypoint/downloads
[preview-downloads-vault]: https://dev-portal-git-zstoast-components-hashicorp.vercel.app/vault/downloads
[preview-swingset]: https://dev-portal-git-zstoast-components-hashicorp.vercel.app/swingset/components/toast
[react-hot-toast]: https://react-hot-toast.com
[@hashicorp/react-toast]: https://github.com/hashicorp/react-components/tree/main/packages/toast
[hashicorp/learn]: https://github.com/hashicorp/learn
[hds-toast]: https://design-system-components-hashicorp.vercel.app/components/toast
[task]: https://app.asana.com/0/1201903760348480/1202114097062999/f
